File: backend/app/schemas/requirement.py
from typing import Optional, List
from datetime import datetime
from .base import SchemaBase

# RequirementVersion Schemas
class RequirementVersionBase(SchemaBase):
    version_string: str
    description: Optional[str] = None
    file_path: Optional[str] = None

class RequirementVersionCreate(RequirementVersionBase):
    requirement_id: int # Required when creating a version

class RequirementVersionUpdate(RequirementVersionBase):
    version_string: Optional[str] = None
    description: Optional[str] = None
    file_path: Optional[str] = None

class RequirementVersion(RequirementVersionBase):
    id: int
    requirement_id: int
    created_at: datetime
    updated_at: datetime
    test_points: List["TestPoint"] = []

    model_config = {"from_attributes": True}


# Requirement Schemas
class RequirementBase(SchemaBase):
    title: str
    # project_id is not in base as it might be part of path parameter for creation
    # or derived from context in some cases.

class RequirementCreate(RequirementBase):
    project_id: int # Explicitly needed for creation

class RequirementUpdate(RequirementBase):
    title: Optional[str] = None
    # project_id is typically not updatable directly, or handled separately.

class Requirement(RequirementBase):
    id: int
    project_id: int
    created_at: datetime
    updated_at: datetime
    versions: List[RequirementVersion] = []

    model_config = {"from_attributes": True}

File: frontend/src/types/requirement.ts
// Based on backend/app/schemas/requirement.py

export interface RequirementVersion {
  id: number; // Integer ID from backend
  requirement_id: number; // Foreign key to Requirement
  version_string: string;
  description?: string | null;
  file_path?: string | null; // Assuming this might be added for file uploads later
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  // test_points?: TestPoint[]; // Example if you expand this type
}

export interface Requirement {
  id: number; // Integer ID from backend
  title: string;
  project_id: number; // Foreign key to Project
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  versions?: RequirementVersion[]; // Usually populated on demand or with specific requests
}

// For creating a new requirement
// Based on backend/app/schemas/requirement.py RequirementCreate
export interface RequirementCreate {
  title: string;
  project_id: number; // Must be provided when creating a requirement
}

// For updating an existing requirement
// Based on backend/app/schemas/requirement.py RequirementUpdate
export interface RequirementUpdate {
  title?: string;
  // project_id is typically not updatable for a requirement, or handled via specific logic
}

// For creating a new requirement version
// Based on backend/app/schemas/requirement.py RequirementVersionCreate
export interface RequirementVersionCreate {
  requirement_id: number; // This is part of the schema, but often also in URL path
  version_string: string;
  description?: string | null;
  file_path?: string | null;
}

// For updating an existing requirement version
// Based on backend/app/schemas/requirement.py RequirementVersionUpdate
export interface RequirementVersionUpdate {
  version_string?: string;
  description?: string | null;
  file_path?: string | null;
}
